File: tree/pathsum/maxPathSum2.java
//from leaf to leaf
public int maxPathSum(TreeNode root){
    int[] max = new int[] {Integer.MIN_VALUE};
    helper(root, max);
    return max[0];
}
//return the value is the max path sum from root to any leaf node in its subtree
private int helper(TreeNode root, int[] max){
    if (root == null)
        return 0;
    int left = helper(root.left, max);
    int right = helper(root.right, max);
    //only when root node has both child, we might update max path sum
    if(root.left != null && root.right != null){
        max[0]=Math.max(max[0], left+right+root.key);
        return Math.max(left, right) +  root.key;
    }
    if(root.left == null)
        return right+root.key;
    else
        return left+root.key
}

File: tree/pathsum/maxPathSum2.py
def maxpathsum2(root):
    if root is None:
        return float('-inf')
    self.max = root.val
    self.dfs(root)
    return self.max
# return single path
def dfs(root):
    if root is None:
        return 0
    left = self.dfs(root.left)
    right = self.dfs(root.right)
    # can not drop negative value, since leaf to leaf

    # has two child (update)
    if root.left and root.right:
        self.max = max(self.max, left+right+root.val)
        return max(left,right) + root.val
    # has only one child
    if not root.left:
        return root.val + right
    else:
        return root.val + left
